ledger/simulation/trace.go

```diff
@@ -67,16 +69,76 @@ func makeTxnGroupResult(txgroup []transactions.SignedTxn) TxnGroupResult {
 // ResultLatestVersion is the latest version of the Result struct
 const ResultLatestVersion = uint64(1)
 
+// ResultEvalConstants contains the limits and parameters during a call to Simulator.Simulate
+type ResultEvalConstants struct {
+	MaxLogCalls *uint64
+	MaxLogSize  *uint64
+}
+
+// ResultEvalConstantsBuilder follows a builder pattern for ResultEvalConstants
+type ResultEvalConstantsBuilder struct {
+	Result ResultEvalConstants
+}
+
+// NewResultEvalConstantsBuilder constructs a builder for ResultEvalConstants
+func NewResultEvalConstantsBuilder() *ResultEvalConstantsBuilder {
+	return &ResultEvalConstantsBuilder{}
+}
+
+// SimulateLogBytesLimit hardcode limit of how much bytes one can log during simulation (with lift-log-limits)
+const SimulateLogBytesLimit = uint64(65536)
+
+// LiftLogLimits method modify the log limits from lift option:
+// - if lift log limits, then overload result from local Config
+// - otherwise, set `LogLimits` field to be nil
+func (r *ResultEvalConstantsBuilder) LiftLogLimits(lift bool) *ResultEvalConstantsBuilder {
+	if lift {
+		maxLogCalls, maxLogSize := uint64(config.MaxLogCalls), SimulateLogBytesLimit
+		r.Result = ResultEvalConstants{
+			MaxLogCalls: &maxLogCalls,
+			MaxLogSize:  &maxLogSize,
+		}
+	}
+	return r
+}
+
+// Finalize method cleanup the *ResultEvalConstants if it is empty ResultEvalConstants{}, then return nil,
+// otherwise it returns the pointer to ResultEvalConstants
+func (r *ResultEvalConstantsBuilder) Finalize() *ResultEvalConstants {
+	// Since ResultEvalConstants is omitempty, we want to check if it is actually empty
+	if r.Result == (ResultEvalConstants{}) {
+		return nil
+	}
+	return &r.Result
+}
+
+// LogicEvalConstants method infers the logic.EvalConstants from Result.EvalConstants (*ResultEvalConstants)
+// and generate appropriate parameters to override during simulation runtime.
+func (c *ResultEvalConstants) LogicEvalConstants() logic.EvalConstants {
+	logicEvalConstants := logic.RuntimeEvalConstants()
+	if c == nil {
+		return logicEvalConstants
+	}
+	if c.MaxLogSize != nil {
+		logicEvalConstants.MaxLogSize = *c.MaxLogSize
+	}
+	if c.MaxLogCalls != nil {
+		logicEvalConstants.MaxLogCalls = *c.MaxLogCalls
+	}
+	return logicEvalConstants
+}
+
 // Result contains the result from a call to Simulator.Simulate
 type Result struct {
-	Version      uint64
-	LastRound    basics.Round
-	TxnGroups    []TxnGroupResult // this is a list so that supporting multiple in the future is not breaking
-	WouldSucceed bool             // true iff no failure message, no missing signatures, and the budget was not exceeded
-	Block        *ledgercore.ValidatedBlock
+	Version       uint64
+	LastRound     basics.Round
+	TxnGroups     []TxnGroupResult // this is a list so that supporting multiple in the future is not breaking
+	WouldSucceed  bool             // true iff no failure message, no missing signatures, and the budget was not exceeded
+	EvalConstants *ResultEvalConstants
```

Comment:
Is there a reason why this is a pointer? Since the fields inside this struct are already pointers, I think that's sufficient information to know whether or they any overrides have happened

As a bonus, if this is no longer a pointer, I think you can remove `ResultEvalConstantsBuilder` and instead have `LiftLogLimits` be a method on `ResultEvalConstants`. As far as I can tell, the reason the builder pattern is needed currently vs directly modifying fields is because `Finalize` will return nil if no overrides have taken place

Reply:
sure, I guess this also implies simplifying logic in `v2/utils.go` for `convertSimulationResult`.

Reply:
see [ResultEvalConstants to instance in Result](https://github.com/algorand/go-algorand/pull/5247/commits/9b9fe08ed03a3d5de19a10ab80b6bac24fccd5c7)